File: perl/recursed/collatzrec.pl
#Mark Gomes
#CSC 330
#Dr. Pounds
#Assignment 3 - Collatz Conjecture

$num = 5000000000;
$count = 0;
$mincount = 0;
@maxcounts = ([0, 0], [0,0], [0,0], [0, 0], [0,0], [0,0], [0, 0], [0,0], [0,0], [0, 0]);
@temp = ([0, 0], [0,0], [0,0], [0, 0], [0,0], [0,0], [0, 0], [0,0], [0,0], [0, 0]);

while ($num > 1) 
{
    $n = $num;
    $count = RecCount($n);      #Preform Collatz algorithm with recursion
    
    #print "Number: $num    Steps: $count \n";
    if(count >= mincount)        #If count is larger than min count in array then add it
    {
        $index = -1;
        $diff = 0;
        $mincount = $count;
   
        for($i = 0; $i < 10; $i++)
        {
            if($maxcounts[$i][1] < $mincount)
            {
                $mincount  = $maxcounts[$i][1];        #Stores the lowest count in the array
            }
            if($count - $maxcounts[$i][1] > $diff)     #Keeps track of index that produces the largest difference between counts
            {
                $diff = $count - $maxcounts[$i][1];
                $index = $i;
            } 
        }
        if($index != -1)
        {
            $maxcounts[$index][1] = $count;
            $maxcounts[$index][0] = $num;
        }
    }
    $num = $num - 1;
    $count = 0;
}

$num = 0;
$count = 0;
for($x = 0; $x < 10; $x++)        #Sort by Step Length
{
    for($i = 0; $i < 10; $i++)
    {
        if($maxcounts[$i][1] >= $count)
        {
            $count = $maxcounts[$i][1];
            $num = $maxcounts[$i][0];
            $index = $i;
        }
    }
    $temp[$x][1] = $count;
    $temp[$x][0] = $num;
    $maxcounts[$index][0] = 0;
    $maxcounts[$index][1] = 0;
    $count = 0;
}
for($i = 0; $i < 10; $i++)
{
    $maxcounts[$i][0] = $temp[$i][0];
    $maxcounts[$i][1] = $temp[$i][1];
}

print "***Array Sorted by Step Length*** \n";
for($i = 0; $i < 10; $i++)        #Display list by Step Length
{
    print "Number: $maxcounts[$i][0]    Steps: $maxcounts[$i][1] \n";
}

$num = 0;
$count = 0;
for($x = 0; $x < 10; $x++)        #Sort by Number Magnitude
{
    for($i = 0; $i < 10; $i++)
    {
        if($maxcounts[$i][0] >= $num)
        {
            $count = $maxcounts[$i][1];
            $num = $maxcounts[$i][0];
            $index = $i;
        }
    }
    $temp[$x][1] = $count;
    $temp[$x][0] = $num;
    $maxcounts[$index][0] = 0;
    $maxcounts[$index][1] = 0;
    $num = 0;
}
for($i = 0; $i < 10; $i++)
{
    $maxcounts[$i][0] = $temp[$i][0];
    $maxcounts[$i][1] = $temp[$i][1];
}

print "***Array Sorted by Number Magnitude*** \n";
for($i = 0; $i < 10; $i++)        #Display list by Number Magnitude
{
    print "Number: $maxcounts[$i][0]    Steps: $maxcounts[$i][1] \n";
}

sub RecCount                  #Recursion to find collatz sequence count
{
    my($n) = @_;
    if($n == 1)
    {
        return 0;
    }
    else
    { 
        if($n%2 == 0)
        {
            return 1 + RecCount($n/2);
        }
        else
        {
            return 1 + RecCount(($n*3)+1);
        }
    }
}

File: java/recursed/collatzrec.java
/* Mark Gomes
 * CSC 330 
 * Dr. Pounds
 * Assignment 3 : Collatz Conjecture
 */

public class collatzrec
{
    public static void main(String[] args)
    {
        long num = 5000000000; 
        int count = 0;
        int mincount = 0;
        
        long[][] maxcounts = new long[10][2];    // counts, number
        
        //Preform Collatz Algorithm
        while(num > 0)
        {
            count = RecCount(num);    //Collatz sequence recursion
            //System.out.print("Number " + num + " has " + count + " steps. \n");
            if(count > mincount)      //If count is larger than min count in array then add it
            {
                UpdateArray(count, num, maxcounts, mincount);
            }
            num = num - 1;
            count = 0;
        }
        
        //Sorted by step length
        System.out.print("***Array Sorted by Step Length*** \n");
        maxcounts = ArraySort(maxcounts, 0);
        for(int i = 0; i < 10; i++)
        {
            System.out.print(i + " : Number " + maxcounts[i][1] + " has " + maxcounts[i][0] + " steps \n");
        }
        
        //Sorted by number magnitude
        System.out.print("***Array Sorted by Number Magnitude*** \n");
        maxcounts = ArraySort(maxcounts, 1);
        for(int i = 0; i < 10; i++)
        {
            System.out.print(i + " : Number " + maxcounts[i][1] + " has " + maxcounts[i][0] + " steps \n");
        }
    }   
    
    static int RecCount(long n)
    {
        if(n == 1)
        {
             return 0;
        }
        else if(n%2 == 0)
        {
            return 1 + RecCount(n/2);
        }
        else
        {
            return 1 + RecCount((n*3)+1);
        }
    }
    
    static void UpdateArray(int count, long num, long[][] maxcounts, int mincount)
    {
        int index = -1;
        int diff = 0;
        mincount = count;
        for(int i = 0; i < 10; i++)
        {
            if((int)maxcounts[i][0] < mincount)
            {
                mincount  = (int)maxcounts[i][0];      //Stores the lowest count in the array
            }
            if(count - (int)maxcounts[i][0] > diff)    //Keeps track of index that produces the largest difference between counts
            {
                diff = count - (int)maxcounts[i][0];
                index = i;
            } 
        }
        if(index != -1)
        {
            maxcounts[index][0] = (long)count;
            maxcounts[index][1] = num;
        }
    }
    
    static long[][] ArraySort(long[][] maxcounts, int mode)    //Mode: [1]Sort by step number; [2]Sort by number magnitude
    {
        long[][] temp = new long[10][2];
        int maxval = 0,count = 0,index = 0;
        long num = 0;
        
        for(int x = 0; x < 10; x++)
        {
            for(int i = 0; i < 10; i++)
            {
                if(maxcounts[i][mode] >= (long)maxval)
                {
                    maxval = (int)maxcounts[i][mode];
                    count = (int)maxcounts[i][0];
                    num = maxcounts[i][1];
                    index = i;
                }
            }
            temp[x][0] = (long)count;
            temp[x][1] = num;
            maxcounts[index][0] = 0;
            maxcounts[index][1] = 0;
            maxval = 0;
        }
        return temp;
    }
}